File: Year-2019/Mar/2019-03-06-240/watercow/240.c
// 从右上角开始(右上角数字是该行最大 该列最小的元素)
// 1）== target: return
// 2) > target: 去掉该数所在列
// 3) < target: 去掉该数所在行
// 4) 直到最后的区域为空
class Solution {
public:
	bool searchMatrix(vector<vector<int>>& matrix, int target) {
		if (matrix.empty()) return false;

		int col = matrix.size(), row = matrix[0].size();
		for (int i = 0, j = row - 1; i < col && j >= 0; ) {
			if (target > matrix[i][j]) i++;
			else if (target < matrix[i][j]) j--;
			else return true;
		}
		return false;
	}
};

File: Year-2019/Mar/2019-03-06-240/lb/240.java
public class SearchMatrix {
    /**
     * 对每一行循环，每次循环修改下次循环的长度的，减少查询的次数
     * 但对于右下角的target需要循环数组的所有数
     * @param matrix
     * @param target
     * @return
     */
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix.length == 0)
            return false;
        int end = matrix[0].length;
        for (int i =0;i<matrix.length;i++){
            for (int j=0;j<end;j++){
                if (matrix[i][j] == target)
                    return true;
                if (matrix[i][j] > target){
                    end = j;//j和之后的数都比target大，修改下次循环的次数
                    break;
                }
            }
        }
        return false;
    }
}
